File: topKFrequent.py
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:

        hashMap = {}
        for n in nums:
            hashMap[n] = 1 + hashMap.get(n,0)

        freqList = []
        for key,value in hashMap.items():
            freqList.append((-value,key))

        heapq.heapify(freqList)

        res = []
        while len(res) < k:
            pop = heapq.heappop(freqList)
            res.append(pop[1])

        return res

File: Top_K_Frequent.java
class Solution {
    //Time O(N)
    //Space O(N)
    public int[] topKFrequent(int[] nums, int k) {
        if(nums == null || nums.length == 0)
        {
            return nums;
        }
        HashMap<Integer , Integer> map = new HashMap<>();
        for(int i = 0 ; i < nums.length ; i++)
        {
            map.put(nums[i] , map.getOrDefault(nums[i] , 0) + 1);
        }
        List bucket[] = new List[nums.length+1];
        for(Integer num : map.keySet())
        {
            int index = map.get(num);
            if(bucket[index] == null)
            {
                bucket[index] = new ArrayList<>();
            }
            bucket[index].add(num);
        }
        int ans[] = new int[k];
        int j = 0;
        for(int i = bucket.length-1 ; i >= 0 && k > 0 ; i--)
        {
            if(bucket[i] != null)
            {
                List<Integer> li = bucket[i];
                for(Integer num : li)
                {
                    ans[j++] = num;
                    k--;
                }
            }
        }
        return ans;
    }
}
